lib/src/retriever.dart

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import "dart:collection";
+
+import "package:pool/pool.dart";
+import "package:async/async.dart";
+
+/// Handles rate-limited scheduling of tasks.
+///
+/// Tasks are named with a key of type [K] (should be useful as a Hash-key) and
+/// run with a supplied function producing a CancelableOperation.
+///
+/// Designed to allow prefetching of tasks that will likely be needed
+/// later with [prefetch].
+///
+/// All current operations can be cancelled and future operations removed from
+/// the queue with [stop].
+///
+/// Errors thrown by tasks scheduled with [prefetch] will only be triggered when
+/// you await the Future returned by [fetch].
+///
+/// The operation will run in the [Zone] that the task was in when enqueued.
+/// If a task if [prefetch]ed and later [fetch]ed before the operation is
+/// started, the task will go in front of the queue with the zone of the [fetch]
+/// operation.
+///
+/// Example:
+///
+/// ```dart
+/// // A retriever that, given a uri, gets that page and returns the body
+/// final retriever = Retriever(
+///     (Uri uri, _) => return CancelableOperation.fromFuture(http.read(uri)));
+/// // Start fetching `pub.dev` in the background.
+/// retriever.prefetch(Uri.parse('https://pub.dev/'));
+/// // ... do time-consuming task.
+///
+/// // Now we actually need `pub.dev`.
+/// final pubDevBody = await retriever.fetch(Uri.parse('https://pub.dev/'));
+/// ```
+class Retriever<K, V> {
+  final CancelableOperation<V> Function(K, Retriever) _run;
```

Comment:
Should this `Retriever` be `Retriever<K,V>`? If not, can we make it explicit that it is a `Retriever<dynamic, dynamic>`?

Reply:
It should indeed be <K,V> thanks!